docs/RFCS/range_repair.md

```diff
@@ -0,0 +1,94 @@
+- Feature Name: Range Repair
+- Status: draft
+- Start Date: 2015-08-19
+- RFC PR:
+- Cockroach Issue: #2149
+
+# Summary
+
+Add a new `Repair` service on each node that monitors all other nodes via
+gossip and determines if any of them are dead. When a node dies, the service
+looks through all the local ranges to see they have any replicas on the dead
+node. If any exist, remove the dead replica and add the local replica to the
+replication queue. See the updated replication queue RFC #2153 that this RFC
+relies on. Note that the `Repair` service is only concerned with remove
+replicas from dead nodes, not adding new ones, that is handled by the
+replication queue.
+
+# Motivation
+
+Repairing the cluster when nodes disappear from the cluster is a base feature
+of cockroach. The goal here is to enable repairing to be quick and effective.
+
+# Detailed design
+
+## Configuration
+Add a new configuration setting called `timeoutUntilNodeDead` which contains
+the number of seconds after which if a node was not heard from, it is
+considered dead. The default value for this will be 5 minutes.
+
+## Repair
+Add a new service called `Repair` that starts when the node is started.
+This new service will run until the stopper is called and have access to both
+gossip and the list of stores (in order to get the list of replicas on those
+stores).
+
+Repair will also maintain a map of node IDs to last update time. To maintain
+this map, a callback from gossip for node descriptors will be added.
+
+Repair will maintain a timespan `timeUntilNextDead` which is calculated by
+taking the nearest last update time for all the nodes and adding
+`timeoutUntilNodeDead` and the node ID associated with the timeout.
+
+Repair will trigger on `timeUntilNextDead` which when triggered checks to see
+if that node has not been updated.
+
+If the node is dead:
+
+1. mark the node as dead
+2. check all the ranges with local replicas to see if any had a replica on this
+   now dead node
```

Comment:
This sounds like a possibly intensive process; should we be maintaining a map of nodes to replicas, or is that just more expensive?